pkg/config/model/viper.go

```diff
@@ -89,13 +90,18 @@ func (c *safeConfig) Set(key string, value interface{}, source Source) {
 		return
 	}
 
-	c.Lock()
-	defer c.Unlock()
-	c.configSources[source].Set(key, value)
-	c.mergeViperInstances(key)
+	// modify the config then release the lock to avoid deadlocks
+	var receivers []NotificationReceiver
+	func() {
+		c.Lock()
+		defer c.Unlock()
+		c.configSources[source].Set(key, value)
+		c.mergeViperInstances(key)
+		receivers = slices.Clone(c.notificationReceivers)
```

Comment:
suggestion: I think this landed in go 1.21, should we use that? 

https://pkg.go.dev/slices#Clone